tests/cairo_files/instructions/test_exchange_operations.cairo

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: MIT
+
+%lang starknet
+
+// Starkware dependencies
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.cairo_builtins import HashBuiltin, BitwiseBuiltin
+from starkware.cairo.common.bool import TRUE, FALSE
+from starkware.cairo.common.uint256 import Uint256, assert_uint256_eq
+from starkware.cairo.common.math import assert_nn
+
+// Local dependencies
+from kakarot.instructions.exchange_operations import ExchangeOperations
+from kakarot.model import model
+from kakarot.stack import Stack
+from tests.utils.utils import TestHelpers
+
+// @notice Prepare a stack with `idx` elements to test swap logic. 
+func prep_stack{range_check_ptr}(idx : felt, to_swap_idx: felt,  stack : model.Stack*) -> (prepped_stack: model.Stack*) {
+    alloc_locals;
+    // We set the last idx to a special number so we can test for successful swapping
+    if (idx == 0) {
+        let updated_stack: model.Stack* = Stack.push(stack, Uint256(2, 0));
+        return (prepped_stack=updated_stack);    
+    }
+    // As well as user defined idx
+    if (to_swap_idx == idx) {
+        let _updated_stack: model.Stack* = Stack.push(stack, Uint256(1, 0));
+        let updated_stack : model.Stack* = prep_stack(idx=idx-1, to_swap_idx=to_swap_idx, stack=_updated_stack);
+        return (prepped_stack=updated_stack);
+    }
+
+    // otherwise we just fill with zero
+    let _updated_stack: model.Stack* = Stack.push(stack, Uint256(0, 0));
+    let updated_stack : model.Stack* = prep_stack(idx=idx-1, to_swap_idx=to_swap_idx, stack=_updated_stack);
+    return (prepped_stack=updated_stack);
+}
+
+// @notice Checks if previously prepared stack has its values properly swapped.
+func check_stack{range_check_ptr}(idx : felt, swapped_idx: felt,  stack : model.Stack*) {
+    alloc_locals;
+    let swapped = Stack.peek(stack, idx);
+    if (idx == 0) {
+        assert swapped = Uint256(1,0);
+        return ();    
+    }
+    // As well as the first
+    if (swapped_idx == idx) {
+        assert swapped = Uint256(2,0);
+    } else {
+        assert swapped = Uint256(0,0);        
```

Comment:
why are you checking all of the stack (with recursion and assert on null values)?

I think here you can make `check_stack` generic, wdyt?
```
func check_swapped_stack{range_check_ptr}(top_stack_element: Uint256, element_at_swapped_idx: Uint256, swapped_idx: felt, stack : model.Stack*) {
   let swapped_element = Stack.peek(stack, 0);
   let swapped_element_at_swapped_idx =  Stack.peek(stack, swapped_idx);
   assert swapped_element = top_stack_element;
   assert swapped_element_at_swapped_idx = element_at_swapped_idx;
```

you'll also need to use uint256_assert_eq (forgot the name) function, it's used in other places of the repo

Reply:
This comment is about: 
1. you don't need to test that the whole stack has been properly filled with zeros in this context, though you could unit test your util `prep_stack`
2. you don't have to tightly couple your two utils (with hardcoded stack values), though you can and it won't change much:)! just that it may decrease readability ❌ but also decrease verbosity ✅
3. you could also design the parameters so that user inputs not the swapped state `top_stack_element` and `element_at_swapped_idx` but the state previous to the swap, something like `preswap_top_stack_element` and `preswap_element_at_swapped_idx`

Reply:
> The thinking was there could be a bug where the logic does swap the two stack positions, but maybe also incorrectly changes an other element of the stack. Checking recursively catches that case, but perhaps is overkill.

I see what you mean, it may not be overkill -> it's always good to check for consistency. But you're testing things outside the scope of this util. 
You'd either have to test prepare_stack unitarily or create a util specifically made for checking the consistency of a stack before/during/after for opcodes that mutate it

Reply:
> I see what you mean, it may not be overkill -> it's always good to check for consistency. But you're testing things outside the scope of this util

Gotcha.

> You'd either have to test prepare_stack unitarily or create a util specifically made for checking the consistency of a stack before/during/after for opcodes that mutate it

Added a unit test for `prepare_stack`.